EventListener/DynamicRelationListener.php

```diff
@@ -0,0 +1,164 @@
+use Doctrine\Common\EventSubscriber;
+use Doctrine\ORM\Event\LoadClassMetadataEventArgs;
+use Doctrine\ORM\Mapping\ClassMetadataInfo;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * A listener that dynamically adds relations for commentable annotation
+ *
+ * @author Nico Kaag <nico.kaag@genj.nl>
+ */
+class DynamicRelationListener implements EventSubscriber
+{
+    /**
+     * @var string $threadClass contains value of 'fos_comment.model.thread.class' parameter
+     */
+    private $threadClass;
+
+    /**
+     * @var Reader
+     */
+    private $reader;
+
+    public function __construct($threadClass, Reader $reader)
+    {
+        $this->threadClass = $threadClass;
+
+        $this->reader = $reader;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function getSubscribedEvents()
+    {
+        return array(
+            'loadClassMetadata',
+            'postLoad',
+        );
+    }
+
+    /**
+     * @param LoadClassMetadataEventArgs $eventArgs
+     */
+    public function loadClassMetadata(LoadClassMetadataEventArgs $eventArgs)
+    {
+        // the $metadata is the whole mapping info for this class
+        $metadata = $eventArgs->getClassMetadata();
+
+        if (!$metadata->getReflectionClass()) {
+            return;
+        }
+
+        $properties = $this->getCommentableProperties($metadata->getReflectionClass());
+
+        if (empty($properties)) {
+            return;
+        }
+
+        $namingStrategy = $eventArgs
+            ->getEntityManager()
+            ->getConfiguration()
+            ->getNamingStrategy()
+        ;
+
+        foreach ($properties as $property) {
+            $metadata->mapOneToOne(
+                array(
+                    'targetEntity' => $this->threadClass,
+                    'fieldName' => $property['property'],
+                    'cascade' => array('persist'),
+                    'fetch' => ClassMetadataInfo::FETCH_LAZY,
+                    'joinColumn' => array(
+                        'name' => 'comment_thread_id',
+                        'referencedColumnName' => 'id'
+                    ),
+                )
+            );
+        }
+    }
+
+    public function postLoad(EventArgs $args)
+    {
+        $this->addCommentThreadToObject($args);
+    }
+
+    private function addCommentThreadToObject(EventArgs $args)
+    {
+        $om = $args->getObjectManager();
+        $object = $args->getObject();
+
+        $reflectionClass = new \ReflectionClass($object);
+
+        if (!$reflectionClass) {
+            return;
+        }
+        $properties = $this->getCommentableProperties($reflectionClass);
+
+        if (empty($properties)) {
+            return;
+        }
+
+        foreach ($properties as $property) {
+            $commentThreadGetter = "get". ucfirst($property['property']);
+            $commentThread = $object->$commentThreadGetter();
+
+            if (!empty($commentThread)) {
+                return;
```

Comment:
should we really return here ? shouln't we rather continue to the next loop iteration ?